### ISSUE TITLE
Posting/removing sub-resources with orphanRemoval break JSON conversion

### ISSUE THREAD

**Issue by Julien-Marcou** [api-platform/api-platform#482]
When I'm saving an entity with a collection of sub-resources who has the following configuration
`@ORM\OneToMany( ... cascade={"persist", "remove"}, orphanRemoval=true)`
Children indexes in the ArrayCollection are broken if orphan children are removed.
Instead of starting at 0, first index start at the first child not removed, which break the JSON conversion, as PHP array can only be converted to JavaScript array, if indexes start at 0 with no gap.

Exemple : 

```php
<?php

namespace AppBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;

/**
 * Parent
 *
 * @ApiResource(
 *     collectionOperations={},
 *     itemOperations={
 *         "put"={
 *             "method"="PUT",
 *             "denormalization_context"={"groups"={"parent-put"}},
 *             "normalization_context"={"groups"={"parent-get"}}
 *         },
 *         "get"={
 *             "method"="GET",
 *             "normalization_context"={"groups"={"parent-get"}}
 *         }
 *     }
 * )
 * @ORM\Table(name="parent")
 */
class Parent
{
    /**
     * @ORM\Id
     * @ORM\Column(name="id", type="integer")
     * @Groups({"parent-get"})
     */
    private $id;

    /**
     * @ORM\OneToMany(targetEntity="Child", mappedBy="parent", cascade={"persist", "remove"}, orphanRemoval=true)
     * @Groups({"parent-get", "parent-put"})
     */
    private $children


    public function __construct()
    {
        $this->children = new ArrayCollection();
    }

    public function addChild(Child $child)
    {
        $child->setParent($this);
        $this->children[] = $child;

        return $this;
    }

    public function removeChild(Child $child)
    {
        $child->setParent(null);
        $this->children->removeElement($child);
    }

    public function getChildren()
    {
        return $this->children;
    }
}
```

```php
<?php

namespace AppBundle\Entity;

/**
 * Child
 *
 * @ORM\Table(name="child")
 */
class Child
{
    /**
     * @ORM\Id
     * @ORM\Column(name="id", type="integer")
     * @Groups({"parent-get"})
     */
    private $id;

    /**
     * @ORM\Column(name="name", type="string", length=255)
     * @Groups({"parent-get", "parent-put"})
     */
    private $name;

    /**
     * @ORM\ManyToOne(targetEntity="Parent", inversedBy="children")
     * @ORM\JoinColumn(name="parent_id", referencedColumnName="id", nullable=false)
     * @Groups({"parent-put"})
     */
    private $parent


    public function setName($name)
    {
        $this->name = $name;
        return $this;
    }

    public function getName()
    {
        return $this->name;
    }

    public function setParent(Parent $parent)
    {
        $this->parent = $parent;
        return $this;
    }

    public function getParent()
    {
        return $this->parent;
    }
}
```

Assuming there is already a Parent entity in the database with the ID 1.
I'm able to put some Child entities to my Parent :
`PUT /api/parent/1`
```json
{
    "children": [
        {
            "name": "riri"
        },
        {
            "name": "fifi"
        }
    ]
}
```
Which returns me something like this
```json
{
    "id": 1,
    "children": [
        {
            "id": 1,
            "name": "riri"
        },
        {
            "id": 2,
            "name": "fifi"
        }
    ]
}
```
So far, so good...

But now, I wan't to remove both Child entities, and add a new one, then I should be able to do :
`PUT /api/parent/1`
```json
{
    "children": [
        {
            "name": "loulou"
        }
    ]
}
```
And I should get (because of orphanRemoval=true)
```json
{
    "id": 1,
    "children": [
        {
            "id": 3,
            "name": "loulou"
        }
    ]
}
```
But what I'm actually getting is
```json
{
    "id": 1,
    "children": {
        "3": {
            "id": 3,
            "name": "loulou"
        }
    }
}
```

The children array is converted to an object as if indexes 0 and 1 had been unset after they were added to the collection (may be by the orphanRemoval=true).

Any clue if it is a bug or if I'm doing something wrong ?

**Comment by soyuka** [comment]
Hi! I had the exact same issue see https://github.com/api-platform/core/pull/1534.

TLDR: 

```
    public function getChildren()
    {
        return $this->children->getValues();
    }
```

**Comment by Julien-Marcou** [comment]
Oh, thanks, that's it !
